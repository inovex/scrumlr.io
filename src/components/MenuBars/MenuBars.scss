@import "constants/style.scss";

.menu,
.menu-bars {
  z-index: $menu-z-index;
}

@media screen and (max-width: 1343px) {
  .menu-bars {
    position: fixed;
    left: calc(50% - 134px);
    bottom: 8px;
    -webkit-touch-callout: none;
  }
  .menu-bars--isAdmin {
    left: calc(50% - 168px);
  }
  .menu-bars--user {
    @media (prefers-color-scheme: light) {
      .menu-bars__switch {
        background-color: $color-white;
        box-shadow: 0 16px 32px 0 rgba(0, 87, 255, 0.24);
      }
      .menu-bars__switch > .switch__icon {
        color: $menu-color-admin--primary;
      }
    }
    @media (prefers-color-scheme: dark) {
      .menu-bars__switch {
        background-color: $menu-color-admin--primary;
        box-shadow: 0 6px 9px 0 rgba(0, 0, 0, 0.2);
      }
      .menu-bars__switch > .switch__icon {
        color: $color-white;
      }
    }
  }
  .menu-bars--user > .user-menu {
    visibility: visible;
    opacity: 1;
    bottom: 8px;
  }
  .menu-bars--user > .admin-menu {
    visibility: hidden;
    opacity: 0;
    bottom: 72px;
  }
  .menu-bars--user > .menu-bars__switch > .switch__icon--add {
    visibility: hidden;
    opacity: 0;
    transform: rotate(360deg);
  }

  .menu-bars--admin {
    @media (prefers-color-scheme: light) {
      .menu-bars__switch {
        background-color: $color-white;
        box-shadow: 0 16px 32px 0 rgba(0, 87, 255, 0.24);
      }
      .menu-bars__switch > .switch__icon {
        color: $menu-color-regular--primary;
      }
    }
    @media (prefers-color-scheme: dark) {
      .menu-bars__switch {
        background-color: $menu-color-regular--primary;
        box-shadow: 0 6px 9px 0 rgba(0, 0, 0, 0.2);
      }
      .menu-bars__switch > .switch__icon {
        color: #ffffff;
      }
    }
  }
  .menu-bars--admin > .admin-menu {
    visibility: visible;
    opacity: 1;
    bottom: 8px;
  }
  .menu-bars--admin > .user-menu {
    visibility: hidden;
    opacity: 0;
    bottom: -56px;
  }
  .menu-bars--admin > .menu-bars__switch > .switch__icon--settings {
    visibility: hidden;
    opacity: 0;
    transform: rotate(-360deg);
  }
  .menu-bars__switch {
    height: 64px;
    width: 64px;
    padding: 0;
    border: 0;
    border-radius: 100%;
    transition: background-color 0.25s ease-in-out;
    margin-left: 272px;
    position: relative;
  }
  .menu-bars__switch:focus {
    outline: none;
  }

  .switch__icon {
    position: absolute;
    top: calc(50% - 18px);
    left: calc(50% - 18px);
    transform: rotate(0deg);
    transition: opacity 0.5s, visibility 0.5s, transform 1s;
  }
}

@media #{$menu-desktop} {
  .menu-bars__switch {
    display: none;
  }
}

.user-menu {
  --primaryColor: #0057ff;
  --secondaryColor: #253f71;
}

.admin-menu {
  --primaryColor: #ff0069;
  --secondaryColor: #7c123e;
}

@media #{$menu-mobile} {
  .menu {
    border-radius: 35px;
    height: 64px;
    overflow: hidden;
    width: 268px;
    visibility: hidden;
    opacity: 0;
    position: fixed;
    margin-right: 64px;
    transition: bottom 0.4s ease-in-out, opacity 0.6s, visibility 0.6s;
  }

  .admin-menu {
    bottom: 72px;
  }

  .user-menu {
    bottom: -56px;
  }

  .menu__items {
    align-items: center;
    box-sizing: border-box;
    display: flex;
    flex-direction: row;
    height: 100%;
    justify-content: space-between;
    padding: 0 28px;
    position: relative;
  }
}

@media #{$menu-desktop} {
  .user-menu {
    left: 8px;
  }

  .admin-menu {
    right: 8px;
  }

  .menu {
    border-radius: 35px;
    height: 268px;
    top: calc(50% - 134px);
    width: 64px;
    overflow: visible;
    position: fixed;
  }

  .menu__items {
    align-items: center;
    box-sizing: border-box;
    display: flex;
    flex-direction: column;
    height: 100%;
    justify-content: space-between;
    padding: 28px 0;
    position: relative;
  }
}

@media (prefers-color-scheme: light) {
  .menu {
    background-color: $color-white;
    box-shadow: 0 16px 32px 0 rgba(0, 87, 255, 0.24);
  }
}

@media (prefers-color-scheme: dark) {
  .menu {
    box-shadow: 0 6px 9px 0 rgba(0, 0, 0, 0.2);
  }

  .user-menu {
    background-color: $menu-color-regular--primary;
  }

  .admin-menu {
    background-color: $menu-color-admin--primary;
  }
}
