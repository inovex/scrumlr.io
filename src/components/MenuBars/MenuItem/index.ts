export * from "./MenuButton";
export * from "./MenuToggle";
