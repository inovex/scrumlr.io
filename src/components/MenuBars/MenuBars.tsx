import {useState} from "react";
import Parse from "parse";
import classNames from "classnames";
import {useSelector} from "react-redux";
import {ApplicationState} from "types/store";
import MenuButton from "components/MenuBars/MenuItem/MenuButton";
import MenuToggle from "components/MenuBars/MenuItem/MenuToggle";
import {ReactComponent as AddImageIcon} from "assets/icon-addimage.svg";
import {ReactComponent as AddStickerIcon} from "assets/icon-addsticker.svg";
import {ReactComponent as CheckIcon} from "assets/icon-check.svg";
import {ReactComponent as SettingsIcon} from "assets/icon-settings.svg";
import {ReactComponent as ColumnIcon} from "assets/icon-column.svg";
import {ReactComponent as FocusIcon} from "assets/icon-focus.svg";
import {ReactComponent as TimerIcon} from "assets/icon-timer.svg";
import {ReactComponent as VoteIcon} from "assets/icon-vote.svg";
import {ReactComponent as ToggleSettingsMenuIcon} from "assets/icon-toggle-settings-menu.svg";
import {ReactComponent as ToggleAddMenuIcon} from "assets/icon-toggle-add-menu.svg";
import "./MenuBars.scss";

function MenuBars() {
  const [showAdminMenu, toggleMenus] = useState(false);

  const currentUser = Parse.User.current();
  const admins = useSelector((state: ApplicationState) => state.users.admins);

  const isAdmin = admins.map((admin) => admin.id).indexOf(currentUser!.id) !== -1;

  return (
    <div className={classNames("menu-bars", {"menu-bars--admin": showAdminMenu, "menu-bars--user": !showAdminMenu}, {"menu-bars--isAdmin": isAdmin})}>
      <div className="menu user-menu">
        <div className="menu__items">
          <MenuToggle direction="right" toggleStartLabel="Mark me as done" toggleStopLabel="Unmark me as done" icon={CheckIcon} onToggle={() => null} />
          <MenuButton direction="right" label="Add image or giphy" icon={AddImageIcon} onClick={() => null} />
          <MenuButton direction="right" label="Add sticker" icon={AddStickerIcon} onClick={() => null} />
          <MenuButton direction="right" label="Settings" icon={SettingsIcon} onClick={() => null} />
        </div>
      </div>
      {isAdmin && (
        <div className="menu admin-menu">
          <div className="menu__items">
            <MenuToggle direction="left" toggleStartLabel="Start column mode" toggleStopLabel="End column mode" icon={ColumnIcon} onToggle={() => null} />
            <MenuToggle direction="left" toggleStartLabel="Start timer" toggleStopLabel="Stop timer" icon={TimerIcon} onToggle={() => null} />
            <MenuToggle direction="left" toggleStartLabel="Start voting phase" toggleStopLabel="End voting phase" icon={VoteIcon} onToggle={() => null} />
            <MenuToggle direction="left" toggleStartLabel="Start focused mode" toggleStopLabel="End focused mode" icon={FocusIcon} onToggle={() => null} />
          </div>
        </div>
      )}
      {isAdmin && (
        <button className="menu-bars__switch" onClick={() => toggleMenus((prevState) => !prevState)}>
          <ToggleAddMenuIcon className="switch__icon switch__icon--add" />
          <ToggleSettingsMenuIcon className="switch__icon switch__icon--settings" />
        </button>
      )}
    </div>
  );
}

export default MenuBars;
