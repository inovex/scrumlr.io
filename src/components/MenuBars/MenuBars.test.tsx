import { render } from "@testing-library/react";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import Parse from 'parse';
import MenuBars from './MenuBars';

const mockStore = configureStore();

const createMenuBars = (state) => {
    const store = mockStore(state)

    return (
        <Provider store={store}>
            <MenuBars />
        </Provider>
    );
};

describe('Menu', () => {
    beforeEach(() => {
        const mockCurrentUser = jest.fn(() => ({id: 'testId'}));
        Parse.User.current = mockCurrentUser;
    });

    test('should render both add- and settings-menu for moderators', () => {
        const state = {
            users: {
                admins: [{id: 'testId'}]
            }
        };
        const {container} = render(createMenuBars(state));
        expect(container.firstChild).toMatchSnapshot();
    });

    test('should only render add-menu for participants', () => {
        const state = {
            users: {
                admins: []
            }
        };
        const {container} = render(createMenuBars(state));
        expect(container.firstChild).toMatchSnapshot();
    })
})