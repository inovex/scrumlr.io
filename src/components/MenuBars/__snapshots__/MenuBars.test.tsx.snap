// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should only render add-menu for participants 1`] = `
<aside
  class="menu-bars menu-bars--user"
  id="menu-bars"
>
  <section
    class="menu user-menu"
  >
    <div
      class="menu__items"
    >
      <button
        class="menu-item menu-item--disabled menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Mark me as done
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-check.svg
        </svg>
        <svg
          class="menu-item__icon menu-item__icon--end"
        >
          icon-close.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Add image or giphy
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-addimage.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Add sticker
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-addsticker.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Settings
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-settings.svg
        </svg>
      </button>
    </div>
  </section>
</aside>
`;

exports[`Menu should render both add- and settings-menu for moderators 1`] = `
<aside
  class="menu-bars menu-bars--user menu-bars--isAdmin"
  id="menu-bars"
>
  <section
    class="menu user-menu"
  >
    <div
      class="menu__items"
    >
      <button
        class="menu-item menu-item--disabled menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Mark me as done
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-check.svg
        </svg>
        <svg
          class="menu-item__icon menu-item__icon--end"
        >
          icon-close.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Add image or giphy
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-addimage.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Add sticker
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-addsticker.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--right"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Settings
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-settings.svg
        </svg>
      </button>
    </div>
  </section>
  <section
    class="menu admin-menu"
  >
    <div
      class="menu__items"
    >
      <button
        class="menu-item menu-item--disabled menu-item--left"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Start column mode
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-column.svg
        </svg>
        <svg
          class="menu-item__icon menu-item__icon--end"
        >
          icon-close.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--disabled menu-item--left"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Start timer
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-timer.svg
        </svg>
        <svg
          class="menu-item__icon menu-item__icon--end"
        >
          icon-close.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--disabled menu-item--left"
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Start voting phase
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-vote.svg
        </svg>
        <svg
          class="menu-item__icon menu-item__icon--end"
        >
          icon-close.svg
        </svg>
      </button>
      <button
        class="menu-item menu-item--disabled menu-item--left"
        disabled=""
      >
        <div
          class="menu-item__tooltip"
        >
          <span
            class="tooltip__text"
          >
            Start focused mode
          </span>
        </div>
        <svg
          class="menu-item__icon menu-item__icon--start"
        >
          icon-focus.svg
        </svg>
        <svg
          class="menu-item__icon menu-item__icon--end"
        >
          icon-close.svg
        </svg>
      </button>
    </div>
  </section>
  <button
    class="menu-bars__switch"
  >
    <svg
      class="switch__icon switch__icon--add"
    >
      icon-toggle-add-menu.svg
    </svg>
    <svg
      class="switch__icon switch__icon--settings"
    >
      icon-toggle-settings-menu.svg
    </svg>
  </button>
</aside>
`;
